src/internal/permissions.ts

```diff
@@ -0,0 +1,418 @@
+
+  type NodeDataResponse = Array<{
+    type: string;
+    data: {
+      perms: {
+        data: Array<{
+          key: { data: string; type_name: string };
+          value: string;
+        }>;
+      };
+    };
+  }>;
+
+  const data = (await res.json()) as NodeDataResponse;
+
+  const permissions = data[0].data.perms.data.map((d) => {
+    switch (d.key.type_name) {
+      case "0x1::fungible_asset::WithdrawPermission": // can this be better? i dont rly like this
+        return FungibleAssetPermission({
+          asset: d.key.data,
+          balance: d.value
+        });
+      case "0x1::object::TransferPermission":
+        return NFTPermission({
+          assetAddress: d.key.data,
+          capabilities: { transfer: true, mutate: false }
+        });
+      default:
+        return GenericPermission({
+          type: d.key.type_name,
+          data: d.key.data,
+          value: d.value
+        });
+    }
+  });
+
+  const filtered = filter ? permissions.filter((p) => filter.includes(p.type as PermissionType)) : permissions;
+  return filtered as FilteredPermissions<T>;
+}
+
+  // should it return the requested permissions? on success? and when it fails it
+export async function requestPermission(args: {
+  aptosConfig: AptosConfig;
+  primaryAccount: SingleKeyAccount;
+  permissionedAccount: AbstractedEd25519Account;
+  permissions: Permission[];
+  expiration?: number;
+  requestsPerSecond?: number;
+}): Promise<SimpleTransaction> {
+  const { aptosConfig, primaryAccount, permissionedAccount, permissions } = args;
+  const aptos = new Aptos(aptosConfig);
+
+  // Get or create a handle for the permissioned account
+  const existingHandleAddress = await getHandleAddress({ aptosConfig, primaryAccount, subAccount: permissionedAccount });
+  
+  return aptos.transaction.build.batched_intents({
+    sender: primaryAccount.accountAddress,
+    builder: async (builder) => {
+      // Get the permissioned signer - either create new one or use existing
+      const permissionedSigner = await getPermissionedSigner(builder, {
+        existingHandleAddress,
+        primaryAccount: primaryAccount,
+        permissionedAccount: permissionedAccount
+      });
+
+      // Grant each requested permission
+      await Promise.all(permissions.map((permission) => 
+        grantPermission(builder, {
+          permissionedSigner: permissionedSigner.signer,
+          permission
+        })
+      ).flat());
+
+      // If we created a new handle, finalize the setup
+      if (permissionedSigner.isNewHandle) {
+        await finalizeNewHandle(builder, {
+          permissionedAccount,
+          handle: permissionedSigner.handle!
+        });
+      }
+
+      return builder;
+    },
+  });
+}
+
+export async function revokePermissions(args: {
+  aptosConfig: AptosConfig;
+  primaryAccount: SingleKeyAccount;
+  subAccount: AbstractedEd25519Account;
+  permissions: RevokePermission[];
```

Comment:
Without putting any thought into this, something feels off here. Need to dig a little.

Reply:
could be a little odd having union of `Permission` but thought process is just passing in the whole object since it has props needed. Depends on how dApps manage this information.